2.9BSD/usr/contrib/grab/man/grab.1
.\"	$Header:   RCS/grab.1.v  Revision 1.2  83/07/01  19:34:59  donn  Exp$
.\"	$Log:	RCS/grab.1.v $
.\"	Revision 1.2  83/07/01  19:34:59  donn
.\"	Added descriptions of '-i' and '-L' options for inode grabbing
.\"	and dumping.
.\"							Donn
.\"	
.\"
.TH GRAB 1 5/18/82
.UC 4
.SH NAME
grab \- get files from filesystems
.SH SYNOPSIS
.B grab
[ -246pxvilLtb ] filesystem [ name ... ]
.SH DESCRIPTION
.PP
.I Grab
is a program which reads Unix filesystem-format devices
and extracts named files.
.I Grab
will run on Version 6 Unix, 2.9 BSD Unix or 4.1 BSD Unix and
will read filesystems made on any of these three.
It is useful for reading filesystems that
cannot be mounted for various reasons
(e.g. there is damage,
system mountable filesystem parameter is set too low,
you are using dual-ported disks or controllers,
the filesystem is not the same type as the one
.I grab
is running on, etc.)
It can also be used as a fast disk copy program,
especially for big chunks of file systems,
since it can restore ownerships and permissions on files,
recreate links and make directories and devices.
.PP
.I Grab
needs to be supplied with the name of a filesystem format file or device,
and a list of pathnames for files on that filesystem
starting at the filesystem root.
For example if you need to extract a file
/mnt/donn/c/quote.c where /mnt is normally mounted
on /dev/xy2, then you type
.br
     grab /dev/xy2 donn/c/quote.c
.br
and this will copy the remote file into a file named
.I quote.c
in your current directory.
(Mounted devices can be listed using
.IR df (1)
or
.IR mount (8).)
If a directory is given as a name,
all the files in that directory are (recursively) extracted
into a subdirectory in the current directory
that has the same name as the remote directory.
If any files extracted are linked together
on the remote filesystem,
the copies
.I grab
makes will also be linked.
The permission information on the files is copied
except for owner, setuid, setgid and sticky information
which must be requested specifically by the
.B \-p
option.
Device nodes will be copied for the super-user.
Here are the options in detail:
.TP 8
.B  -2
The remote filesystem is a 2.9 BSD filesystem.
(Note: By default the remote filesystem
is assumed to be the same type as the host system.)
.TP 8
.B -4
The remote filesystem is a 4.1 BSD filesystem.
.TP 8
.B -6
The remote filesystem is a Version 6 filesystem.
(Note: On version 6 grab cannot consistently read filesystems
made on Version 7 that use more than 65K small blocks
since the V6 I/O system uses 16 bits to store block numbers...)
.TP 8
.B  -p
Restore owner and group, setuid, setgid and sticky bits,
and access and modification times.
Only the superuser can reliably copy owner and group.
.I Grab
will not allow protected files on remote filesystems to be read
except by the super-user.
.TP 8
.B  -v
Verbose option -- the names of copied files
are announced as they are finished being copied.
.TP 8
.B  -x
Print option -- the requested files are copied
to the standard output.
If the standard output is a file then
this is the default.
.TP 8
.B  -i
Inode option -- the arguments must be inode numbers
and the program looks at the files indicated by the given inodes.
If files are created without a name being available,
the inode number is converted into a name.
This is useful for recovering files on screwed-up filesystems
when the inode numbers are known through (say)
.IR dcheck (8).
.TP 8
.B  -l
List option -- the names of files in the requested directories are listed.
This is useful if you don't know or can't remember
the name of a file to extract.
.TP 8
.B  -L
Long listing option -- the program lists files individually
and displays useful information about them:
their name, inode number, mode word, link count, uid and gid,
access and modify times, and within-inode block numbers.
.TP 8
.B  -t
.I Tar
option -- the requested files are placed on the standard output
in a format identical to that of the
.IR tar (1)
program.
Digits following the letter
.B t
are taken to be a blocking factor for the tape;
the default blocksize is 20.
To put a Version 6 directory on tape you might type:
.br
   grab -6tv /dev/rxp0d /etc > /dev/rmt0
.TP 8
.B -b
This option allows you to set the size of the buffer
.I grab
uses for reading and writing.
Digits following the
.B b
are taken to be a buffer size in 512-byte blocks.
The default is 128 blocks on the VAX, 32 on PDPs.
Increasing the buffer size generally increases
the speed but you have to be careful not to exceed
the bounds of your machine.
.SH BUGS
Checking uids of remote files is
rather silly if the remote system
is not the same as the host system.
.sp
No warning is given about clobbering pre-existing files.
.sp
Version 6 is obsolete.


2.9BSD/usr/contrib/grab/src/grab.c
/*
 * grab.c
 *
 * Name:	grab
 * Purpose:	extract files from remote filesystems
 * Usage:	grab [-246pxvilLtb#] <filesystem> <files...>
 * Environment:	V6 or V7 (2.9BSD, 4.1BSD) Unix
 * Compile:	cc -O -s -D<type> grab.c find.c readi.c bread.c tmode.c -o grab
 * Date:	4/29/82
 * Author:	Donn Seeley RRCF UCSD
 * Remarks:
 *	This is meant to be used with dual-ported disks or controllers
 *	when it is either impossible or undesirable to mount another
 *	file system on that medium.  It should allow any Version 6 or
 *	Version 7 (2.9BSD, 4.1BSD) system to read filesystems from any
 *	Version 6 or Version 7 (2.9BSD, 4.1BSD) system even if the systems
 *	are not of the same type.  Version 6 systems will need at least
 *	a Phototypesetter level C compiler to make "grab".  "Grab" will
 *	copy any conceivable (?) object on a filesystem including directories
 *	(recursively), links, file holes, setuid/setgid/sticky files and
 *	device nodes.  "Grab" tries to prevent unauthorized readers from
 *	gazing at remote files but the only reliable way of maintaining
 *	security is to make "grab" setgid and then make fs's readable by
 *	group and not by other.  ("Df" should work this way too.)
 *
 *	To compile a V6 grab, define PDPV6.  To compile a 2.9 grab define
 *	PDP, and to compile a 4.1 grab define VAX.  If your machine has
 *	multiplexed files, define MPXFILES, and if it has symbolic links
 *	you should define SYMLINKS.
 *
 *	Options:
 *		-246	2.9, 4.1 or V6 system on the remote end.
 *		-p	Preserve all permissions and ownerships, instead of
 *			just some of them (should be su).
 *		-x	Act like "cat(1)" from the remote fs (but still
 *			recursively on directories).
 *		-v	Verbose mode: print filenames as they are copied.
 *		-i	The arguments are inode numbers; treat them as
 *			files with those inode numbers and named by the numbers.
 *		-l	Act like a rudimentary "ls(1)" on a remote fs.
 *		-L	Long version of -l: gives a gruesome inode dump.
 *		-t	Act like "tar(1)" on a remote fs.  Output goes to
 *			the standard output.  Digits following the "t" are
 *			interpreted as a blocksize.
 *		-b	Set blocksize independently: this manipulates the
 *			internal write buffer size.  Digits following, etc.
 *
 *---------------------------------------------------------------------------
 *
 * $Header:   RCS/grab.c.v  Revision 1.2  83/07/01  19:29:13  donn  Exp$
 * $Log:	RCS/grab.c.v $
 * Revision 1.2  83/07/01  19:29:13  donn
 * Added -i and -L flags for inode grabbing and inode dumping.
 * 							Donn
 * 
 */

# include	"grab.h"

int	target_system	= TARGET_DEFAULT;
int	fsbsiz		= FSB_DEFAULT;
int	fsys;
int	pflag		= 0;
int	xflag		= 0;
int	vflag		= 0;
int	iflag		= 0;
int	lflag		= 0;
int	Lflag		= 0;
int	tflag		= 0;
int	nblock		= B_BUFSIZE;




main( argc, argv )
    int	    argc;
    char   *argv[];
{
	char   *tail();
	char   *name;
	int	ino;

	if ( argc < 3 ) {
		fprintf( stderr, "grab: too few arguments\n" );
		fprintf( stderr, "Usage: grab [-246pxvilLtb] <file-system> <files>\n" );
		exit( 1 );
	}

	/*
	 * Process the command line.
	 */
	while ( *argv[1] == '-' ) {

		while ( *++argv[1] != '\0' )
			switch ( *argv[1] ) {

			/*
			 * Select type of filesystem to read.
			 */
			case '2':
				target_system	= V7_2BSD;
				fsbsiz		= 1024;
				break;
			case '4':
				target_system	= V7_4BSD;
				fsbsiz		= 1024;
				break;
			case '6':
				target_system	= V6;
				fsbsiz		= 512;
				break;

			/*
			 * Restore previous owner, group & protections.
			 */
			case 'p':
				++pflag;
# ifndef PDPV6
				umask( 0 );
# endif
				break;

			/*
			 * Put files on the standard output.
			 */
			case 'x':
				++xflag;
				break;

			/*
			 * Print names of copied files.
			 */
			case 'v':
				++vflag;
				break;

			/*
			 * Grab by inode number.
			 */
			case 'i':
				++iflag;
				break;

			/*
			 * List directories.
			 */
			case 'l':
				/*
				 * If only one file then we leave out header.
				 */
				lflag	= argc - 3;
				Lflag	= 0;
				break;

			/*
			 * Long listing.
			 */
			case 'L':
				++Lflag;
				lflag	= 0;
				break;

			/*
			 * Output files in "tar" format.
			 */
			case 't':
				++tflag;
				nblock	= TAR_BS_DEFAULT;
				/* FALL THRU */

			/*
			 * Pick up blocking factor (if any).
			 */
			case 'b':
				if ( *(argv[1] + 1) >= '0' &&
				     *(argv[1] + 1) <= '9' )
					nblock	= atoi( ++argv[1] );
				while ( *(argv[1]+1) >= '0' && *(argv[1]+1) <= '9' )
					++argv[1];

				break;

			default:
				fprintf( stderr, "grab: unknown option -%c\n", argv[1][1] );
				fprintf( stderr, "Usage: grab [-246pxvilLtb] <file-system> <files>\n" );
				exit( 2 );
			}
		++argv;
		--argc;
	}

	if ( argc < 3 ) {
		fprintf( stderr, "grab: too few arguments\n" );
		fprintf( stderr, "Usage: grab [-246pxvilLtb] <file-system> <files>\n" );
		exit( 3 );
	}



	/*
	 * If the standard output is a file, assume -x by default.
	 * This is for backward compatibility with Arthur Olsen's grab
	 * and may go away.
	 */
	if ( ! (xflag || tflag || lflag || Lflag) ) {
		int	statbuf[32];

		if ( fstat( STDOUT, statbuf ) == 0
# ifdef PDPV6
		 && (((struct inode *) statbuf)->i_mode & V6_IFMT) == V6_IFREG )
# else
		 && (((struct inode *) statbuf)->i_mode & V7_IFMT) == V7_IFREG )
# endif
			++xflag;
	}



	/*
	 * Try to separate file output and verbose mode output!
	 * The method here is a bit nonportable...
	 */
	if ( tflag || xflag )
		fileno( stdout )	= fileno( stderr );



	/*
	 * Open the file system device.
	 */
	fsys = open( argv[1], 0 );
	if ( fsys < 0 ) {
		fprintf( stderr, "grab: can't open %s\n", argv[1] );
		exit( 4 );
	}



	/*
	 * Set id to real id.  (V6 isn't setuid because nobody cares.)
	 */
# ifndef PDPV6
	setuid( getuid() );
	setgid( getgid() );
# endif



	/*
	 * Find the files corresponding to the file arguments
	 * and read them off.
	 */
	while ( argc > 2 ) {
		if ( iflag ) {
			name	= NULL;
			ino	= atoi( argv[2] );
		} else {
			name	= tail( argv[2] );
			ino	= find( argv[2] );
		}
		readi( ino, name );
		++argv, --argc;
	}



	/*
	 * Clean up on tape output.
	 * Tar puts out two blank records followed by junk
	 * from the buffer when it flushes the buffer...
	 */
	if ( tflag )
		tflush( TFILE );

	exit( 0 );
}
